ncm-spma/src/main/perl/spma/apt.pm

```diff
@@ -0,0 +1,385 @@
+
+=head1 RESOURCES
+
+Only a very minimal schema is implemented.
+
+Sources listed under C</software/repositories> will be configured,
+URLs should be followed by the suite and sections required e.g. http://example.org/debian unstable main
+
+Packages listed under C</software/packages> will be installed, version and architecture locking (including multiarch) is fully implemented.
+
+=cut
+
+use strict;
+use warnings;
+use parent qw(NCM::Component CAF::Path);
+use CAF::Process;
+use CAF::FileWriter;
+use CAF::FileEditor;
+use EDG::WP4::CCM::Path qw(escape unescape);
+use EDG::WP4::CCM::TextRender;
+use NCM::Component::spma::yum;
+use CAF::Object qw(SUCCESS);
+use Set::Scalar;
+use File::Path qw(mkpath);
+use Text::Glob qw(match_glob);
+use Readonly;
+
+Readonly::Scalar my $DIR_SOURCES => "/etc/apt/sources.list.d";
+Readonly::Scalar my $DIR_PREFERENCES => "/etc/apt/preferences.d";
+Readonly::Scalar my $TEMPLATE_SOURCES => "apt/source.tt";
+Readonly::Scalar my $TEMPLATE_PREFERENCES => "apt/preferences.tt";
+Readonly::Scalar my $TREE_SOURCES => "/software/sourcesitories";
+Readonly::Scalar my $TREE_PKGS => "/software/packages";
+Readonly::Scalar my $BIN_APT_GET => "/usr/bin/apt-get";
+Readonly::Scalar my $BIN_APT_MARK => "/usr/bin/apt-mark";
+Readonly::Scalar my $BIN_DPKG_QUERY => "/usr/bin/dpkg-query";
+Readonly::Scalar my $CMD_APT_UPDATE => [$BIN_APT_GET, qw(-qq update)];
+Readonly::Scalar my $CMD_APT_UPGRADE => [$BIN_APT_GET, qw(-qq dist-upgrade)];
+Readonly::Scalar my $CMD_APT_INSTALL => [$BIN_APT_GET, qw(-qq install)];
+Readonly::Scalar my $CMD_APT_AUTOREMOVE => [$BIN_APT_GET, qw(-qq autoremove)];
+Readonly::Scalar my $CMD_APT_MARK => [$BIN_APT_MARK, qw(-qq)];
+Readonly::Scalar my $CMD_DPKG_QUERY => [$BIN_DPKG_QUERY, qw(-W -f=${db:Status-Abbrev};${Package}\n)];
+
+our $NoActionSupported = 1;
+
+
+# If user specified sources (userrepos) are not allowed, removes any
+# sources present in the system that are not listed in $allowed_sources.
+sub cleanup_old_sources
+{
+    my ($self, $sources_dir, $allowed_sources) = @_;
+    $self->debug(5, 'Entered cleanup_old_sources()');
+
+    my $current = Set::Scalar->new(glob("$sources_dir/*.list"));
+    my $allowed = Set::Scalar->new(map($_->{name}, @$allowed_sources));
+    my $to_remove = $current-$allowed;
+
+    if (-d $sources_dir) {
+        foreach my $s (@$to_remove) {
+            my $fn = "$sources_dir/$s.list";
+            $self->verbose("Unlinking outdated source $fn");
+            if (!defined (CAF::Path->cleanup($fn, ""))) {
+                return 0;
+            }
+        }
+    } else {
+        $self->error("Cannot cleanup sources, $sources_dir doesn't exist.");
+        return 0;
+    }
+    return 1;
+}
+
+
+# Creates the sources directory if needed.
+sub initialize_sources_dir
+{
+    my ($self, $sources_dir) = @_;
+    $self->debug(5, 'Entered initialize_sources_dir()');
+
+    if (CAF::Path->directory($sources_dir) != $sources_dir) {
```

Comment:
this is not what you want. for string comparison you need `ne`, but on failure, this will give a warning, since `->directory` retruns undef.
you can do
```perl
if (! $self->directory($sources_dir)) {
    $self->error("Unable to create source dir $sources_dir: $self->{fail}");
    return 0;
}
```